concordium-smart-contract-testing/src/types.rs

```diff
@@ -360,6 +378,55 @@ impl ContractInvokeSuccess {
     }
 }
 
+/// A wrapper for [`ContractTraceElement`], which provides additional
+/// information for testing and debugging. Most notably, it contains trace
+/// elements for failures, which are normally discarded by the node.
+#[derive(Debug)]
+pub enum DebugTraceElement {
+    /// A regular trace element with some additional data, e.g., energy usage
+    /// and the entrypoint.
+    /// This variant may be included included in the `WithFailures` list of
+    /// trace elements.
+    Regular {
+        /// The entrypoint.
+        entrypoint:    OwnedEntrypointName,
+        /// The trace element.
+        trace_element: ContractTraceElement,
+        /// The energy used so far.
+        energy_used:   Energy,
+    },
+    /// One or multiple trace elements that fail. Useful for debugging.
+    /// This variant also contains additional information, such as the error,
+    /// entrypoint, and energy usage.
+    WithFailures {
+        /// The address of the contract which failed.
+        contract_address: ContractAddress,
+        /// The entrypoint which failed.
+        entrypoint:       OwnedEntrypointName,
```

Comment:
Yes, that is true. They can be seen from the last Regular in the `trace_elements`. So we can remove the fields from the `WithFailures` variant. 
But for the `Regular`, the `entrypoint` is not redundant, and it is actually very useful in my experience.

For the test_case_1, with the flow
```
/// A
///  -->  B
///         --> A
///         <--
///       B(trap)
/// A <--
```
The following trace is produced:

<details>
  <summary>Click me to view the output</summary>
 
```
[
    Regular {
        entrypoint: OwnedEntrypointName(
            "a_modify_proxy",
        ),
        trace_element: Interrupted {
            address: ContractAddress {
                index: 0,
                subindex: 0,
            },
            events: [],
        },
        energy_used: Energy {
            energy: 616,
        },
    },
    WithFailures {
        contract_address: ContractAddress {
            index: 1,
            subindex: 0,
        },
        entrypoint: OwnedEntrypointName(
            "b_forward_crash",
        ),
        error: Trap {
            error: ExecutionError(
                "Unreachable.",
            ),
        },
        trace_elements: [
            Regular {
                entrypoint: OwnedEntrypointName(
                    "b_forward_crash",
                ),
                trace_element: Interrupted {
                    address: ContractAddress {
                        index: 1,
                        subindex: 0,
                    },
                    events: [],
                },
                energy_used: Energy {
                    energy: 944,
                },
            },
            Regular {
                entrypoint: OwnedEntrypointName(
                    "a_modify",
                ),
                trace_element: Updated {
                    data: InstanceUpdatedEvent {
                        contract_version: V1,
                        address: ContractAddress {
                            index: 0,
                            subindex: 0,
                        },
                        instigator: Contract(
                            ContractAddress {
                                index: 1,
                                subindex: 0,
                            },
                        ),
                        amount: Amount {
                            micro_ccd: 0,
                        },
                        message: OwnedParameter(
                            [],
                        ),
                        receive_name: OwnedReceiveName(
                            "a.a_modify",
                        ),
                        events: [],
                    },
                },
                energy_used: Energy {
                    energy: 1273,
                },
            },
            Regular {
                entrypoint: OwnedEntrypointName(
                    "b_forward_crash",
                ),
                trace_element: Resumed {
                    address: ContractAddress {
                        index: 1,
                        subindex: 0,
                    },
                    success: true,
                },
                energy_used: Energy {
                    energy: 1273,
                },
            },
        ],
        energy_used: Energy {
            energy: 1273,
        },
    },
    Regular {
        entrypoint: OwnedEntrypointName(
            "a_modify_proxy",
        ),
        trace_element: Resumed {
            address: ContractAddress {
                index: 0,
                subindex: 0,
            },
            success: false,
        },
        energy_used: Energy {
            energy: 1273,
        },
    },
    Regular {
        entrypoint: OwnedEntrypointName(
            "a_modify_proxy",
        ),
        trace_element: Updated {
            data: InstanceUpdatedEvent {
                contract_version: V1,
                address: ContractAddress {
                    index: 0,
                    subindex: 0,
                },
                instigator: Account(
                    AccountAddress(
                        [
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                            0,
                        ],
                    ),
                ),
                amount: Amount {
                    micro_ccd: 1,
                },
                message: OwnedParameter(
                    [
                        1,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        36,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        8,
                        0,
                        97,
                        95,
                        109,
                        111,
                        100,
                        105,
                        102,
                        121,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        15,
                        0,
                        98,
                        95,
                        102,
                        111,
                        114,
                        119,
                        97,
                        114,
                        100,
                        95,
                        99,
                        114,
                        97,
                        115,
                        104,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                        0,
                    ],
                ),
                receive_name: OwnedReceiveName(
                    "a.a_modify_proxy",
                ),
                events: [],
            },
        },
        energy_used: Energy {
            energy: 1274,
        },
    },
]
```
</details>

Reply:
I see the convenience it adds now, so I am fine leaving it as is

Reply:
We discussed this a bit in the office before this was implemented, and without having these explicit, it is actually quite annoying to recover them since you have to go back from the end.

So I think it's good to have them. However we should document a bit more that they are redundant.

Reply:
As per your other comment, @abizjak, I've now included a `WithFailures` element even when there are no trace elements. In those cases, the contract and entrypoint aren't redundant, but I've added a comment which implies redundancy when the `trace_elements` vector has elements.